geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LPushDUnitTest.java

```diff
@@ -0,0 +1,216 @@
+  private static MemberVM locator;
+
+  @ClassRule
+  public static RedisClusterStartupRule clusterStartUp = new RedisClusterStartupRule();
+
+  @Rule
+  public ExecutorServiceRule executor = new ExecutorServiceRule();
+
+  private static final int MINIMUM_ITERATIONS = 10000;
+  private static JedisCluster jedis;
+
+  @BeforeClass
+  public static void classSetup() {
+    locator = clusterStartUp.startLocatorVM(0);
+    clusterStartUp.startRedisVM(2, locator.getPort());
+    clusterStartUp.startRedisVM(3, locator.getPort());
+  }
+
+  @Before
+  public void testSetup() {
+    clusterStartUp.startRedisVM(1, locator.getPort());
+    clusterStartUp.rebalanceAllRegions();
+    int redisServerPort = clusterStartUp.getRedisPort(1);
+    jedis = new JedisCluster(new HostAndPort(BIND_ADDRESS, redisServerPort), REDIS_CLIENT_TIMEOUT);
+    clusterStartUp.flushAll();
+  }
+
+  @AfterClass
+  public static void tearDown() {
+    jedis.close();
+  }
+
+  @Test
+  public void shouldPushMultipleElementsAtomically()
+      throws ExecutionException, InterruptedException {
+    AtomicLong runningCount = new AtomicLong(PUSHER_COUNT);
+
+    List<String> listHashtags = makeListHashtags();
+    List<String> keys = makeListKeys(listHashtags);
+    List<String> elements1 = makeElementList(PUSH_LIST_SIZE, "element1-");
+    List<String> elements2 = makeElementList(PUSH_LIST_SIZE, "element2-");
+    List<String> elements3 = makeElementList(PUSH_LIST_SIZE, "element3-");
+
+    List<Runnable> taskList = new ArrayList<>();
+    taskList.add(() -> lpushPerformAndVerify(keys.get(0), elements1, runningCount));
+    taskList.add(() -> lpushPerformAndVerify(keys.get(0), elements1, runningCount));
+    taskList.add(() -> lpushPerformAndVerify(keys.get(1), elements2, runningCount));
+    taskList.add(() -> lpushPerformAndVerify(keys.get(1), elements2, runningCount));
+    taskList.add(() -> lpushPerformAndVerify(keys.get(2), elements3, runningCount));
+    taskList.add(() -> lpushPerformAndVerify(keys.get(2), elements3, runningCount));
+    taskList.add(() -> verifyListLengthCondition(keys.get(0), runningCount));
+    taskList.add(() -> verifyListLengthCondition(keys.get(1), runningCount));
+    taskList.add(() -> verifyListLengthCondition(keys.get(2), runningCount));
+
+    List<Future> futureList = new ArrayList<>();
+    for (Runnable task : taskList) {
+      futureList.add(executor.runAsync(task));
+    }
+
+    for (int i = 0; i < 50 && runningCount.get() > 0; i++) {
+      clusterStartUp.moveBucketForKey(listHashtags.get(i % listHashtags.size()));
+      GeodeAwaitility.await().during(Duration.ofMillis(500)).until(() -> true);
+    }
+
+    for (Future future : futureList) {
+      future.get();
+    }
+
+    int totalLength = 0;
+    Long length;
+    for (String key : keys) {
+      length = jedis.llen(key);
+      assertThat(length).isGreaterThanOrEqualTo(MINIMUM_ITERATIONS * 2 * PUSH_LIST_SIZE);
+      totalLength += length;
+    }
```

Comment:
It's not clear to me how this test is verifying that list elements are added atomically, since it doesn't verify the contents of the lists, just their lengths. We could have a situation where we do two LPUSH operations with elements `1, 1, 1` and `a, a, a`, and end up with a list `1, 1, a, a, a, 1` and the test wouldn't mind. Would it be possible to add validation that every three consecutive elements of the list correspond to one single LPUSH operation? To make it easier, you could push three identical but distinct elements each time, like the example I gave, then just check that every three list elements are equal to each other to make sure nothing got applied out of order.

Reply:
Without something like LINDEX that kind of validation can only be done destructively. There absolutely is a failure mode where only part of a list gets added if it's not handled atomically. We can add a ticket to update these with a LINDEX.

Reply:
I think it's fine to validate destructively at this point, because we've stopped doing any operations on the keys and don't need to do any other validation once we confirm the contents of the lists. I agree that it would be good to update the tests once we have LINDEX, but there's nothing to stop us from adding that extra confirmation that the behaviour is correct right now.